src/stm32f0xx/gpio.c

```diff
@@ -0,0 +1,76 @@
+#include "gpio.h"
+
+#include "stm32f0xx.h"
+#include "stm32f0xx_gpio.h"
+
+static GPIO_TypeDef *gpio_port_map[] = {GPIOA, GPIOB, GPIOC, GPIOD, GPIOE, GPIOF};
+
+// Map the uint8_t version of the pin to the uint16_t used by the gpio functions.
+// This isn't a linear map hence the switch case.
+static uint16_t prv_pin_map(uint8_t pin) {
+  assert_param(pin < 16);
+  return 0x01 << pin;
+}
+
+// Parse the GPIO settings and a pin number to create an initialization stuct to be used by the
+// CMSIS HAL.
+static GPIO_InitTypeDef *prv_parse_args(uint8_t pin, GPIOSettings *settings) {
+  GPIO_InitTypeDef init_struct;
```

Comment:
Static doesn't guarantee inlining. You were returning a pointer to an object you'd allocated on the stack, so it would've returned an address to invalid memory once the function returned.